neverwinter/nwscript/compilerapi.h

```diff
@@ -0,0 +1,127 @@
+#pragma once
+
+//
+// To build the shared library on MacOS:
+//   g++ -O2 -fPIC -shared -std=c++14 \
+//     -o libscriptcomp.dylib \
+//     neverwinter/nwscript/compilerapi.cpp \
+//     neverwinter/nwscript/native/*.{cpp,c}
+//
+
+#include <cstdint>
+#include <stddef.h>
+#include <stdint.h>
+
+typedef uint16_t RESTYPE;
+typedef uint32_t STRREF;
+
+//
+// High level flow:
+// 1) Check ABI version.
+// 2) Create a compiler instance, give it callbacks into your space. Make sure to also call init.
+// 3) Compile a file.
+// 4) Service all callbacks - serve files and collect artifacts.
+// 5) Receive the compilation artifacts in the write callback.
+// 6) Compile more or destroy the compiler instance when done.
+//
+
+//
+// Called by the compiler when it wants to write a file.
+// Return 0 on success, or an error STRREF on failure.
+//
+typedef int32_t (*ResManWriteToFile)(const char* sFileName, RESTYPE nResType, const uint8_t* pData,
+    size_t nSize, bool bBinary);
+
+//
+// Called by the compiler when it wants the contents of a script file.
+//
+// Call Compiler->DeliverFile(...) to satisfy the request during the invocation
+// of this callback. The convoluted to/from is to work around managed memory
+// tracking issues on some languages.
```

Comment:
Ugh. Can you please give an example of where the current flow is failing? I can't imagine why this is better than just returning it here, somehow.

Reply:
```
def handle_cb(...) -> bytes: #cdecl
  return b"somestr"
```

Python (and other managed languages) allocate "somestr" on the heap and adds it to GC, but has no way to track the lifecycle. It returns it into unmanaged code and has no idea when it'll be no longer needed. It cannot GC before returning, obviously, and instead of GC'ing it the next time python gets called it just leaks it. 

I tried it with the return pattern and python was very noisy about this. I think this is some kind of GC contract python establishes because there might not be a GIL in the future and the GC might recycle it on a side thread while native code is still running.

Reply:
Thanks. My care-o-meter here is really really low, so just do whatever, but in this case I think what I would do is something like:
```
def handle_cb(...) -> bytes: #cdecl 
    x = b"somestr"
    cleanup_on_compile_finish.append(x)
    return x
```

Reply:
I tried that in python by adding it to a global dict, but it still moaned about it. It's probably just a hardcoded warning. I don't like this now any more than you do, but I assumed other languages might be similarly iffy about losing allocated data to native code. So I just gave up and accepted that passing byte buffers directly into return values is bad form.

Reply:
I guess that'd also work, but it'd need a way to get the raw bytes and write it into a pointer from the managed side. I'm not immediately sure how to do that in python, but I'm sure it can be done. Mind, I want this to also not only target python but other languages, so keeping the interface as simple as possible.

It's different now because the argument never leaves python space untracked: the call to DeliverFile happens and python regains control, where it can hold on to the object and GC it at the end of the function/callback.